### ISSUE TITLE
Convention (or something) to force static dependency to Sinks assemblies loaded dynamically

### ISSUE THREAD

**Issue by caioproiete** [serilog/serilog#880]
(ideas ... to be discussed).

---

This is not a new issue (it has always been like that), but it is more apparent now that every sink is in its own assembly.

When you have Sinks that are loaded dynamically, for example, via App.config, e.g.:

```
    Log.Logger = new LoggerConfiguration()
        .ReadFrom.AppSettings()
        .CreateLogger();

    <?xml version="1.0" encoding="utf-8" ?>
    <configuration>
      <appSettings>
        <add key="serilog:minimum-level" value="Verbose" />

        <add key="serilog:using:RollingFile" value="Serilog.Sinks.RollingFile" />
        <add key="serilog:write-to:RollingFile.restrictedToMinimumLevel" value="Verbose" />
        <add key="serilog:write-to:RollingFile.pathFormat" value="..\log\myapp-{Date}.log" />
      </appSettings>
    </configuration>
```

There is no static link between the app and the sinks' assemblies and its dependencies (in this case, `Serilog.Sinks.File`, and `Serilog.Sinks.RollingFile`).

This means that if your log configuration lives on `ClassLibrary1`, which is referenced by `ConsoleApp1`, the assemblies for the sinks will **not** be copied to the bin folder of `ConsoleApp1`.

Also, even if the log configuration is defined in `ConsoleApp1`, tools like ReSharper will tell you (rightfully so) that it is safe to remove references to these assemblies, in whatever project you installed the NuGet packages, so you have to remember that every time you're doing a clean-up.

![image](https://cloud.githubusercontent.com/assets/177608/19449751/3c932836-947d-11e6-94f4-42c90efdeb44.png)

---

Couple of ideas that come to mind, which could solve this issue:

**1.) Add a convention in which each Sink exposes an extension method such as `UseSinkName` or `RequireSinkName`**, that developers can optionally call when configuring the logger. e.g.

```
    Log.Logger = new LoggerConfiguration()
        .UsingRollingFileSink()
        .UsingSeqSink()
        .ReadFrom.AppSettings()
        .CreateLogger();
```

These `UsingSinkName` methods, of course, would be in each of the Sinks' assemblies, and therefore create the static dependency. They wouldn't do anything other than return the same `LoggerConfiguration` passed in.

**2.) Add more granular support to read settings for each Sink**, so as an alternative to using `.ReadFrom.AppSettings()` to read everything at once, we could have an overload, by Sink, to indicate the settings come from somewhere else. E.g.

```
    Log.Logger = new LoggerConfiguration()
        .WriteTo.RollingFile(c => c.ReadFrom.AppSettings)
        .WriteTo.Seq(c => c.ReadFrom.KeyValuePairs(...))
        .CreateLogger();
```

---

Thoughts?

**Comment by khellang** [comment]
Ah, the good old linking referenced, but "unused" assemblies feature. Lovely. Reminded me of Nancy's (`IncludeInNancyAssemblyScanningAttribute`)(https://github.com/NancyFx/Nancy/blob/master/src/Nancy/IncludeInNancyAssemblyScanningAttribute.cs) :stuck_out_tongue_closed_eyes:

**Comment by caioproiete** [comment]
@khellang Exactly! 😝  - And jokes aside, an assembly attribute per sink would also be a valid way to solve this issue.

Curiously, the workaround I'm using today is similar to MvvmCross' [LinkerPleaseInclude class](https://github.com/MvvmCross/NPlus1DaysOfMvvmCross/blob/master/N-25-Tabbed/Tabbed.Droid/LinkerPleaseInclude.cs) that you mentioned on one of the PRs for Nancy.

**Comment by nblumhardt** [comment]
Interesting problem to try to tackle elegantly :-)

It feels like it's a general project/build system issue rather than something that's specific to Serilog, since a lot of reflection-based code suffers from this.

A cheap-and-nasty fix we could add with minimum fuss is:

```csharp
[assembly: ImplicitDependency(typeof(RollingFileLoggerConfigurationExtensions))]
```

I.e., stamped into the top-level "program" this would just embed the already-exposed extension class into some metadata.

Interestingly, I don't think .NET Core apps have this issue - IIRC all dependencies are carried over even if they're not used directly. If that's the case, keeping this off of the `LoggerConfiguration` API seems like the right thing to do.

Attribute:

```csharp
namespace Serilog.Settings
{
    [AttributeUsage(AttributeTargets.Assembly)]
    public sealed class ImplicitDependencyAttribute : Attribute
    {
        public Type ImplicitlyUsedType { get; }

        public ImplicitDependencyAttribute(Type implicitlyUsedType)
        {
            ImplicitlyUsedType = implicitlyUsedType;
        }
    }
}
```

?

**Comment by caioproiete** [comment]
@nblumhardt Yeah, you convinced me. It really is more of a build system issue rather than a Serilog-only issue, and a generic assembly attribute like you suggest seems like a good approach and definitely cheap.

I initially thought about a generic extension method on the `LoggerConfiguration` API e.g. which would also be cheap.
```
new LoggerConfiguration().Using<RollingFileLoggerConfigurationExtensions>()
```

... instead of the (more expensive) specific extension methods inside each sink which I suggested on # 1 above, but I didn't like how it read, as the type referenced is not a really a "sink".

In retrospect, expecting every sink author to implement an extension method doesn't seem reasonable.

Anyway, unless anyone have a better idea, I'd say :shipit: for the `ImplicitDependencyAttribute`

**Comment by nblumhardt** [comment]
Sounds good to me. Perhaps so that it's a bit less surprising to encounter, we could call it `[ConfigurationDependency]` or `[SettingsDependency]`?

**Comment by caioproiete** [comment]
Oh, now we arrived at the really hard things... Naming :)

Both are good to me, really. I'll give a vote to `[ConfigurationDependency]` just because it reminds me that it is a dependency of my `LoggerConfiguration` (in fact, `[LoggerConfigurationDependency]` is not bad either).

**Comment by caioproiete** [comment]
**TL;DR;** Maybe we should keep the name `ImplicitDependencyAttribute`? Or maybe `SerilogDependencyAttribute`?

---

Long version:

OK, so I was about to send a PR for this, to get things going, and did a couple of tests before just to double-check it works fine, and I noticed that this issue is not exclusive to assemblies that are loaded dynamically. It also applies to other assemblies such as sink dependencies, that are not directly referenced (in code) by the app.

Example:
- File -> New Project
- `Install-Package Serilog`
- `Install-Package Serilog.Sinks.Seq`
- Setup log config:
```
    Log.Logger = new LoggerConfiguration()
        .WriteTo.Seq("https://seq.mycompany.com")
        .CreateLogger();
```
![image](https://cloud.githubusercontent.com/assets/177608/20121035/3f43f170-a5c4-11e6-8ddf-4b31a6f3ce06.png)

With that said, maybe we should keep the name `ImplicitDependencyAttribute` to be more generic? Or maybe `SerilogDependencyAttribute`?

**Comment by nblumhardt** [comment]
I don't think the effect of removing those dependencies will be the same as removing the root (sink) dependency - the assemblies should still be included in the output location because they're referenced transitively, and the compiler usually respects that. Or am I missing a twist? Cheers!

**Comment by caioproiete** [comment]
> I don't think the effect of removing those dependencies will be the same as removing the root (sink) dependency - the assemblies should still be included in the output location because they're referenced transitively, and the compiler usually respects that. Or am I missing a twist? Cheers!

Unfortunately the assemblies don't get copied unless the reference is there.

Quick-and-dirty sample here:
https://github.com/caioproiete/serilog/tree/test-transitive-dependencies

Entry point is `TestTransitiveDependencies` which references (via project reference) project `MyLoggerConfig` that uses the `Serilog` and `Seq` sink, which itself depends on the `PeriodicBatching` and `RollingFile` sinks, which depends on the `File` sink.

The `bin` folder of `TestTransitiveDependencies` only copies `Serilog.dll` and `Serilog.Sinks.Seq.dll`

**Comment by nblumhardt** [comment]
Ah, interesting. Thanks for the follow-up.

It looks like this may be an issue with ReSharper? Found: https://resharper-support.jetbrains.com/hc/en-us/community/posts/206354699-Remove-unused-references-removes-references-that-are-dependencies-of-used-nuget-packages and (consistent with the issue, though apparently deprecated): https://github.com/JetBrains/resharper-nuget/issues/3

**Comment by caioproiete** [comment]
Thanks @nblumhardt.

It seems that the plugin you referenced seems to be something that existed before ReSharper incorporated the "Remove Unused References" feature in the main product, and then after that they deprecated the plugin as no longer needed. My team uses latest ReSharper (2016.2.2 September 13) with zero plugins, and the issue is there.

But regardless, it is definitely an issue with ReSharper not detecting the dependencies properly and also an issue that we don't get any MSBuild warnings when assemblies that are being referenced, are not there... I'd expect to get a compile-time error, but nope...

**Comment by nblumhardt** [comment]
Thanks for the additional info 👍 my guess is that the bug/limitation got carried across from the plug-in to the product.

I'm going to walk through the "used from XML configuration" side of this scenario to make sure it all hangs together; if by shipping this we can work around the R# issue I think that's a good outcome, but I would rather hold back on adding this if it's only useful in that context.

**Comment by skomis-mm** [comment]
I believe since [2016.1](https://blog.jetbrains.com/dotnet/2016/04/14/resharper-ultimate-2016-1-is-released/) R# should handle implicit assembly references better when they are included as nuget-dependencies:
> Optimize References learns to handle NuGet references, and Remove Unused References starts to support NuGet references as well.

So, in @caioproiete example I see this:
![image](https://cloud.githubusercontent.com/assets/15090984/20338958/0c41ddbe-abeb-11e6-8bc5-5c8e1be83ca9.png)

**Comment by nblumhardt** [comment]
Howdy all - since the ReSharper update mentioned by Sergey above, is this still a pain point?

**Comment by caioproiete** [comment]
Hey, I confirm the bug is fixed with most recent version of ReSharper.
